### DISCUSSION TITLE
Question for block.timestamp

### DISCUSSION THREAD

**Discussion by KittyFu307** [zkSync-Community-Hub/zksync-developers#13] [General] [answered]
### Environment

Mainnet

### Provide a brief description of the functionality you're trying to implement.

I am trying to figure out the timestamp update of zksync era.
Have a few question of block.timestamp.
1. Assuming that the proof has not been generated yet, but tx has not been sent to L1, how do I get this timestamp? 
2. Will there be a certain delay between block.timestamp and executed time?
3. Is it possible that the transaction is confirmed on l2 but fails to be submitted to l1?
4. Is the timestamp of l1 also used when the transaction is executed on l2?
5.How often will ZkSync get the timestamp of L1?

### What is the specific issue or error you're encountering?

I am trying to log the timestamp of ZkSync Era. And I send two tx on the mainet. the two tx were log the timestamp of my tx. But there are in different block but return the same block.timestamp which comfused me.

### Can you share the error messages or logs you're receiving, if any?

<img width="1087" alt="img" src="https://github.com/zkSync-Community-Hub/zkync-developers/assets/61280608/4d80588f-8e14-43dc-bdaa-af2c26f67147">

### Have you made any recent changes to the contract before encountering this issue?

no

### Are there any external libraries or contracts that your contract interacts with?

https://explorer.zksync.io/address/0xbbdF04a4c418f09Bab41a6d74AbDC39339df9772#events

### Can you provide the relevant portions of your contract code where the issue is occurring?

no

### Have you tried to isolate the problem, and if so, what were the results?

no

### What steps have you already taken to try to resolve the issue?

no

**Comment by bxpana** [comment]
Here's our docs that have more details about how block number and timestamp work on zkSync Era for now https://era.zksync.io/docs/dev/developer-guides/transactions/blocks.html#block-number-and-timestamp-considerations

In the future we will be working on implementing so it returns the L2 block timestamp

**Comment by KittyFu307** [comment]
Yeah I have already read the the doc but still have above questions. @bxpana could you give me an explanation for those questions? How often will ZkSync ERA get L1’s timestamp?

**Comment by KittyFu307** [comment]
@bxpana and have a more question 
Is timestamp updated when  batching transaction? How often when batching transactions and how to update the timestamps

**Comment by uF4No** [comment, marked as answer]
block.timestamp will return the timestamp from the latest batch produed. In your case, your transactions were included in the batch [#50922](https://explorer.zksync.io/batch/50922) which was produced at the same timestamp that your transactions emit in the event, 1685714192 (Fri Jun 02 2023 13:56:32 GMT+0000). Hope that helps understand how it currently works.

A batch will include transactions from multiple blocks and the batch is sealed/produced based on different parameters like timeout or just space in the batch, so there's not a fixed time to produce batches. You can read about that in [this page of the docs](https://era.zksync.io/docs/dev/developer-guides/transactions/blocks.html#processing-times). The more usage in the network, the faster batches will be sealed/produced.

> Will there be a certain delay between block.timestamp and executed time?

Yes, for the moment block.timestamp will have a delayed value as it returns the timestamp the last produced batch. 

> Is it possible that the transaction is confirmed on l2 but fails to be submitted to l1?

Very unlikely, only if the proof cannot be generated.

> Is the timestamp of l1 also used when the transaction is executed on l2?

Not exactly of L1, but from the latest batch produced.

Finally, we're working on different methods to allow developers to access both the L2 timestamp and the L1 timestamp.

**Reply by bxpana** [reply]
@KittyFu307

**Reply by Sam-Scolari** [reply]
By "last produced batch" do you mean the latest sealed batch or validated batch? I am trying to figure out if calling block.number in the EVM will give me the latest L1 block number that is either roughly a few minutes old or a day old due to finality times.

**Comment by KittyFu307** [comment]
Thank you so much!

**Comment by guarinojoe** [comment]
Timestamps in zkSync Era are typically obtained from Layer 1 (L1) Ethereum blockchain. However, there may be delays between the timestamp reported by zkSync and the actual execution time due to various factors like network congestion. It's possible for transactions to be confirmed on Layer 2 (L2) but fail to be submitted to L1. Timestamps on L1 are generally used when transactions are executed on L2. The frequency of timestamp updates from L1 to zkSync may vary and is typically determined by zkSync's architecture and implementation. In your case, receiving the same block timestamp for transactions in different blocks could be due to zkSync's batching mechanism or other internal processes. Further investigation may be needed to understand the exact behavior.

**Comment by kimballthomas** [comment]
It's important to understand that block.timestamp in zkSync is different from Ethereum's block.timestamp. In zkSync, the timestamp is determined by the prover, which is responsible for creating and submitting proofs to Ethereum. Here are some key points to consider:

Timestamp Before Proof Generation: If the proof has not been generated yet, the timestamp may not be available immediately. It depends on when the prover generates the proof and submits it to Ethereum.

Delay Between block.timestamp and Execution Time: There might be a delay between the block.timestamp and the actual execution time of the transaction, especially if there are network congestion or other factors affecting transaction processing.

Confirmation on L2 vs Submission to L1: It's possible for a transaction to be confirmed on L2 but fail to be submitted to L1 if there are issues with the bridge or other related components.

Timestamp Usage: In zkSync, the timestamp is used internally for various purposes, but it's not directly tied to the timestamp on L1 Ethereum.

Timestamp Update Frequency: The timestamp of L1 Ethereum is periodically updated by zkSync. However, the exact frequency may vary and is determined by zkSync's protocol and network conditions.

Regarding your observation of receiving the same block.timestamp for transactions in different blocks, it's possible that the prover submitted the proofs for those transactions at around the same time, resulting in similar timestamps. Additionally, zkSync's timestamp resolution might not be fine-grained enough to distinguish between transactions submitted close together.

**Comment by hanjv** [comment]
The timestamp in zkSync reflects the time when the transaction is included in a zkSync block, not when it's executed on L1. There might be some delay between block.timestamp and the actual execution time. Transactions can fail to be submitted to L1 even after being confirmed on L2. Timestamps in L1 and L2 are independent. zkSync updates its timestamp from L1 periodically.

**Comment by maestaspat** [comment]
🕒 Timestamps in zkSync Era can be accessed through block.timestamp, but there might be nuances to consider:

1. Timestamps are obtained from L1 Ethereum when a transaction is executed on L2.
2. There could be a slight delay between block.timestamp and the actual execution time due to network factors.
3. Transactions can fail to be submitted to L1 even after confirmation on L2, but this is rare.
4. The timestamp from L1 is utilized when transactions are executed on L2.
5. Timestamp updates in zkSync are synced with L1 Ethereum periodically.

Regarding your specific issue, logging the timestamp of transactions in different blocks returning the same block.timestamp might indicate a synchronization or caching issue.

**Comment by sifrediblue** [comment]
The user is inquiring about the timestamp update in the zkSync Era network on the Mainnet environment. They have several questions regarding the block.timestamp, including how to get the timestamp if the proof has not been generated yet, whether there is a delay between block.timestamp and executed time, and whether the timestamp of L1 is used when a transaction is executed on L2. The user has encountered an issue where two transactions logged the same block.timestamp even though they were in different blocks. They have not made any recent changes to the contract and have not tried to isolate the problem or resolve the issue yet. Further investigation and debugging may be needed to determine the cause of the issue and the feasibility and impact of any necessary changes.
